File: src/core/formats.py
"""Image format detection and supported extension registry."""

from enum import Enum, auto
from pathlib import Path


class FormatGroup(Enum):
    PILLOW = auto()  # JPEG, PNG, WebP, GIF, BMP, TIFF
    HEIF = auto()    # HEIC, HEIF, AVIF (via pillow-heif)
    RAW = auto()     # CR2, NEF, ARW, DNG, etc. (via rawpy)
    SVG = auto()     # SVG (via librsvg / Gdk.Texture)


_FORMAT_MAP: dict[str, FormatGroup] = {}

_PILLOW_EXTS = {".jpg", ".jpeg", ".png", ".webp", ".gif", ".bmp", ".tif", ".tiff"}
_HEIF_EXTS = {".heic", ".heif", ".avif"}
_RAW_EXTS = {".cr2", ".cr3", ".nef", ".nrf", ".arw", ".dng", ".orf", ".raf", ".rw2"}
_SVG_EXTS = {".svg"}

for ext in _PILLOW_EXTS:
    _FORMAT_MAP[ext] = FormatGroup.PILLOW
for ext in _HEIF_EXTS:
    _FORMAT_MAP[ext] = FormatGroup.HEIF
for ext in _RAW_EXTS:
    _FORMAT_MAP[ext] = FormatGroup.RAW
for ext in _SVG_EXTS:
    _FORMAT_MAP[ext] = FormatGroup.SVG

SUPPORTED_EXTENSIONS: frozenset[str] = frozenset(_FORMAT_MAP.keys())


def is_supported(filename: str) -> bool:
    """Check if a filename has a supported image extension."""
    return Path(filename).suffix.lower() in SUPPORTED_EXTENSIONS


def detect_format(filename: str) -> FormatGroup | None:
    """Detect the format group for a filename. Returns None if unsupported."""
    return _FORMAT_MAP.get(Path(filename).suffix.lower())

File: src/core/formats.rs
use std::path::Path;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FormatGroup {
    /// Raster formats we can decode to a bitmap
    Image,
    /// SVG — loaded via gdk4::Texture::from_filename (librsvg)
    Svg,
}

const RASTER_EXTENSIONS: &[&str] = &[
    "jpg", "jpeg", "jpe", "jfif",
    "png", "webp", "gif", "bmp",
    "tif", "tiff",
    "avif", "heic", "heif",
    "jxl",
    "ico", "tga", "qoi", "hdr", "dds",
    "pbm", "pgm", "ppm", "pnm",
    "ff", "exr",
];

fn extension(filename: &str) -> String {
    Path::new(filename)
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| e.to_ascii_lowercase())
        .unwrap_or_default()
}

/// Check if a filename has a supported image extension.
pub fn is_supported(filename: &str) -> bool {
    detect_format(filename).is_some()
}

/// Detect the format group for a filename. Returns None if unsupported.
pub fn detect_format(filename: &str) -> Option<FormatGroup> {
    let ext = extension(filename);
    if ext == "svg" {
        return Some(FormatGroup::Svg);
    }
    if RASTER_EXTENSIONS.contains(&ext.as_str()) {
        return Some(FormatGroup::Image);
    }
    None
}
